security/src/main/java/com/fincity/security/service/UserService.java

```diff
@@ -400,5 +415,212 @@ public Mono<Boolean> removingPermissionFromUser(ULong userId, ULong permissionId
+
+		        userId -> checkHierarchy(userId, reqUserId, requestPassword.getNewPassword()),
+
+		        (userId, isUpdatable) -> isUpdatable.booleanValue() ? reqClientId : Mono.empty(),
+
+		        (userId, isUpdatable, clientId) -> clientId != null
+		                ? this.clientPasswordPolicyService.getPolicyByClientId(clientId)
+		                : Mono.empty(),
+
+		        (userId, isUpdatable, clientId, passwordPolicy) -> passwordPolicy != null
+		                ? this.clientPasswordPolicyService.checkAllConditions(clientId,
+		                        requestPassword.getNewPassword())
+		                : Mono.just(true),
+
+		        (userId, isUpdatable, clientId, passwordPolicy, isValid) ->
+
+				isValid.booleanValue()
+				        ? this.addBasedOnHistoryCount(reqUserId, requestPassword.getNewPassword(), userId,
+				                passwordPolicy)
+				        : Mono.empty(),
+
+		        (userId, isUpdatable, clientId, passwordPolicy, isValid,
+		                addPastPassword) -> addPastPassword.booleanValue()
+		                        ? this.dao.updatePassword(reqUserId, requestPassword.getNewPassword())
+		                        : Mono.empty(),
+
+		        (userId, isUpdatable, clientId, passwordPolicy, isValid, addPastPassword, passwordUpdated) ->
+				{
+
+			        if (passwordUpdated.booleanValue()) {
+				        this.soxLogService.create(new SoxLog().setObjectId(userId)
+				                .setActionName(SecuritySoxLogActionName.OTHER)
+				                .setObjectName(SecuritySoxLogObjectName.USER)
+				                .setDescription("Password updated"))
+				                .subscribe();
+
+				        this.dao.makeUserActive(reqUserId)
+				                .subscribe(); // making user active if it is another state
+
+				        return Mono.just(true);
+			        }
+
+			        return Mono.empty();
+
+		        }
+
+		).log()
+
+		        .switchIfEmpty(securityMessageResourceService.throwMessage(HttpStatus.FORBIDDEN,
+		                "Password cannot be updated"));
+
+	}
+
+	// add past password check here with dao and service here
+
+	public Mono<UShort> getHistoryCount(ULong clientId) {
+
+		return flatMapMono(
+
+		        () -> clientService.getClientPasswordPolicy(clientId),
+
+		        passwordPolicy -> passwordPolicy != null && passwordPolicy.getPassHistoryCount() != null
+		                ? Mono.just(passwordPolicy.getPassHistoryCount())
+		                : Mono.just(UShort.valueOf(0)));
+
+	}
+
+	// delete past password
+
+	public Mono<List<String>> fetchList(ULong userId) {
+		return this.dao.fetchPasswordsById(userId);
+	}
+
+	// add new password based on history count which was applicable for that
+	// selected user id
+
+	public Mono<Boolean> addBasedOnHistoryCount(ULong reqUserId, String password, ULong lUserId,
+	        ClientPasswordPolicy passwordPolicy) {
+
+		return flatMapMono(
+
+		        () -> Mono.just(reqUserId),
+
+		        userId -> this.dao.fetchPasswordsCount(userId),
```

Comment:
Usually, you don't count the passwords. You must insert it into past passwords and then write a delete query based on the client password policy. It has to be executed separately by calling subscribe method.

Reply:
removed this method from add past password method.